packages/devtools_app/lib/src/charts/treemap.dart

```diff
@@ -0,0 +1,517 @@
+import '../utils.dart';
+
+class Treemap extends StatefulWidget {
+  const Treemap({
+    @required this.rootNode,
+    @required this.levelsVisible,
+    @required this.width,
+    @required this.height,
+    @required this.onTap,
+  });
+
+  final TreemapNode rootNode;
+
+  /// The depth of children visible from this Treemap widget.
+  ///
+  /// For example, levelsVisible = 2 at root Treemap:
+  /// ```
+  /// _______________
+  /// |     Root    |
+  /// ---------------
+  /// |    l = 1    |
+  /// |  ---------  |
+  /// |  | l = 2 |  |
+  /// |  |       |  |
+  /// |  |       |  |
+  /// |  ---------  |
+  /// ---------------
+  /// ```
+  final int levelsVisible;
+
+  final double width;
+
+  final double height;
+
+  final VoidCallback onTap;
+
+  @override
+  _TreemapState createState() => _TreemapState();
+}
+
+enum PivotType { pivotByMiddle, pivotBySize }
+
+class _TreemapState extends State<Treemap> {
+  static const double minHeightToDisplayText = 20.0;
+  PivotType pivotType = PivotType.pivotBySize;
+
+  TreemapNode rootNode;
+
+  @override
+  void initState() {
+    super.initState();
+    rootNode = widget.rootNode;
+  }
+
+  bool get shouldDisplayText => widget.height > minHeightToDisplayText;
+
+  void cellOnTap(TreemapNode child) {
+    if (child != null) {
+      setState(() {
+        rootNode = child;
+      });
+    }
+  }
+
+  /// Computes the total size of a given list of treemap nodes.
+  int computeByteSizeForNodes({
+    @required List<TreemapNode> nodes,
+    @required int startIndex,
+    @required int endIndex,
+  }) {
+    int sum = 0;
+    for (int i = startIndex; i <= endIndex; i++) {
+      sum += nodes[i].byteSize;
+    }
+    return sum;
+  }
+
+  int computePivot(List<TreemapNode> children) {
+    switch (pivotType) {
+      case PivotType.pivotByMiddle:
+        return (children.length / 2).floor();
+      case PivotType.pivotBySize:
+        int pivotIndex = -1;
+        double maxSize = double.negativeInfinity;
+        for (int i = 0; i < children.length; i++) {
+          if (children[i].byteSize > maxSize) {
+            maxSize = children[i].byteSize.toDouble();
+            pivotIndex = i;
+          }
+        }
+        return pivotIndex;
+      default:
+        return -1;
+    }
+  }
+
+  /// Divides a given list of treemap nodes into four parts:
+  /// L1, P, L2, L3.
+
+  /// P (pivot) is the treemap node chosen to be the pivot based on the pivot type.
+  /// L1 includes all treemap nodes before the pivot treemap node.
+  /// L2 and L3 combined include all treemap nodes after the pivot treemap node.
+  /// A combination of elements are put into L2 and L3 so that
+  /// the aspect ratio of the pivot cell (P) is as close to 1 as it can be.
+
+  /// Example layout:
+  /// ```
+  /// ----------------------
+  /// |      |  P   |      |
+  /// |      |      |      |
+  /// |  L1  |------|  L3  |
+  /// |      |  L2  |      |
+  /// |      |      |      |
+  /// ----------------------
+  /// ```
```

Comment:
can you add a link here to the algorithm source this code is based on?

Reply:
Done